### DISCUSSION TITLE
Debug portable_pty spawning programs with Ok(ExitStatus { code: 3221225794, signal: None }), which in hex is 0xc0000142 == STATUS_DLL_INIT_FAILED

### DISCUSSION THREAD

**Discussion by michaelfortunato** [wezterm/wezterm#4674] [Q&A]
### What Operating System(s) are you running on?

Windows

### WezTerm version

n/a

### Ask your question!

This question is about the `protable_pty` Rust crate.
I am able to launch processes fine using this crate on UNIX like systems, but every time I try to spawn a program using
```rust
           let mut cmd =
            portable_pty::CommandBuilder::new("cmd.exe");

        let pair = pty_system.openpty(pty_size.into())?;
        debug!(
            "Creating pty for client: ({}) With child process: {}",
            client_addr,
            self.cmd.as_unix_command_line().unwrap()
        );
        let child = pair.slave.spawn_command(self.cmd.clone())?;
       child.wait()
```
`child.wait()` returns immediately with exit code Ok(ExitStatus { code: 3221225794, signal: None }), which in hex is 0xc0000142 == STATUS_DLL_INIT_FAILED. _This happens with any command I put in, not just `cmd.exe`_. Has anyone else had this issue?

**Comment by wez** [comment]
I'd suggest googling for "conpty STATUS_DLL_INIT_FAILED"

**Comment by michaelfortunato** [comment]
What I find on my Windows 11 Systems (using Parallels) is that I get STATUS_DLL_INIT_FAILED
if I return the `Box<dyn Child + Send + Sync>` from a function and wait on it in the calling block. On the other hand, if I spawn the child process and call `.wait()` on it in the same block, everything works as expected. Here is the code from the main.rs file reproducing the issue I have. I also made a repo people can clone which reproduces it: https://github.com/isc-mfortuna/portable_pty_conpty_bug
```rust
use portable_pty::{native_pty_system, Child, CommandBuilder, PtySize};
use std::{io::Read, io::Write};

fn main() {
    let (mut child, _reader, _writer) = spawn_dll_error();
    println!("Child Process Return: {:?}", child.wait());
    /* // Spawning and waiting on the child process in the same  block fixes the issue
        let pty_system = native_pty_system();
        // Create a new pty
        let mut pair = pty_system
            .openpty(PtySize {
                rows: 24,
                cols: 80,
                // Not all systems support pixel_width, pixel_height,
                // but it is good practice to set it to something
                // that matches the size of the selected font.  That
                // is more complex than can be shown here in this
                // brief example though!
                pixel_width: 0,
                pixel_height: 0,
            })
            .unwrap();
        let mut cmd = CommandBuilder::new("cmd");
        let mut child = pair.slave.spawn_command(cmd).unwrap();
        let mut reader = pair.master.try_clone_reader().unwrap();
        let mut writer = pair.master.take_writer().unwrap();
        (child, reader, writer)
        println!("Child Process Return: {:?}", child.wait());
    }
    */
}
fn spawn_dll_error() -> (
    Box<dyn Child + Send + Sync>,
    Box<dyn Read + Send>,
    Box<dyn Write + Send>,
) {
    let pty_system = native_pty_system();
    // Create a new pty
    let mut pair = pty_system
        .openpty(PtySize {
            rows: 24,
            cols: 80,
            // Not all systems support pixel_width, pixel_height,
            // but it is good practice to set it to something
            // that matches the size of the selected font.  That
            // is more complex than can be shown here in this
            // brief example though!
            pixel_width: 0,
            pixel_height: 0,
        })
        .unwrap();
    let mut cmd = CommandBuilder::new("cmd");
    let mut child = pair.slave.spawn_command(cmd).unwrap();
    let mut reader = pair.master.try_clone_reader().unwrap();
    let mut writer = pair.master.take_writer().unwrap();
    (child, reader, writer)
}
```

**Reply by wez** [reply]
I suspect that this is an issue because you are dropping the reader and writer very early, and windows doesn't like it when the handles are closed very quickly during process startup.  You might try something like this:

```rust
fn main() {
    let (mut child, reader, writer) = spawn_dll_error();
    println!("Child Process Return: {:?}", child.wait());
    drop(reader);
    drop(writer);
}
```

but note that you probably should at least try to read data from the reader, and that you may actually need to drop the writer before your `child.wait` to avoid deadlock in some cases (eg: the command won't terminate until it receives EOF, but that won't get sent while the writer is alive)

I recommend putting the child wait in a separate thread from reading and writing.

**Reply by michaelfortunato** [reply]
It turns out it was actually because I was dropping the `pair` created by `pty_system::openpty`. In spawn_dll_error, it does not return the pair so that struct is not moved to the calling function, and instead the rust compiler frees it. What I had to do is keep `pair` around in the IO infinite copy loop so that it stays in scope. Then everything works fine.
